// angular-material-prefix-updater/material/component-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** Export the class name data as part of a module. This means that the data is cached. */
exports.classNames = require('./data/class-names.json');
/** Export the input names data as part of a module. This means that the data is cached. */
exports.inputNames = require('./data/input-names.json');
/** Export the element selectors data as part of a module. This means that the data is cached. */
exports.elementSelectors = require('./data/element-selectors.json');
/** Export the attribute selectors data as part of a module. This means that the data is cached. */
exports.exportAsNames = require('./data/export-as-names.json');
/** Export the attribute selectors data as part of a module. This means that the data is cached. */
exports.attributeSelectors = require('./data/attribute-selectors.json');
/** Export the property names as part of a module. This means that the data is cached. */
exports.propertyNames = require('./data/property-names.json');
/** Method that can be used to remove the surrounding Angular attribute brackets from a string. */
exports.removeAttributeBackets = function (str) { return str.substring(1, str.length - 1); };
